### ISSUE TITLE
a Pre-training problem

### ISSUE THREAD

**Issue by zjuyb** [mkshing/e4t-diffusion#7]
Thanks for the code. When I use the provided code and the provided pre-training script, when I step 15348/100000, the loss gets nan。

accelerate launch pretrain_e4t.py \
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" \
  --clip_model_name_or_path="openai/clip-vit-large-patch14" \
  --domain_class_token="art" \
  --placeholder_token="*s" \
  --prompt_template="a photo of {placeholder_token}" \
  --reg_lambda=0.01 \
  --output_dir="pretrained-art" \
  --train_image_dataset="Artificio/WikiArt" \
  --iterable_dataset \
  --resolution=512 \
  --train_batch_size=16 \
  --learning_rate=1.6e-5 \
  --max_train_steps=100_000 \
  --mixed_precision="fp16" \
  --enable_xformers_memory_efficient_attention \
  --use_8bit_adam

`Steps:  15%|█▌        | 15348/100000 [9:48:58<44:40:40,  1.90s/it, train/loss=nan, train/loss_diff=nan, train/loss_reg=nan, train/lr=1.6e-5]

so what happened？what should I do？     `

**Comment by mkshing** [comment]
I believe this issue https://github.com/d8ahazard/sd_dreambooth_extension/issues/859 is related and I think you installed xformers==0.0.16 and you need to upgrade xformers such as 

```
pip install xformers==0.0.17.dev466
```

**Comment by dong-y21** [comment]
> 

I updated xformers, but it doesn't work for me.
Still get NaN while training.

**Comment by zjuyb** [comment]
> I believe this issue [d8ahazard/sd_dreambooth_extension#859](https://github.com/d8ahazard/sd_dreambooth_extension/issues/859) is related and I think you installed xformers==0.0.16 and you need to upgrade xformers such as
> 
> ```
> pip install xformers==0.0.17.dev466
> ```
Could not find a version that satisfies the requirement xformers==0.0.17.dev466
So I updated xformers=0.0.17.dev480, but it doesn't work for me.
Still get NaN while training.

**Comment by laolongboy** [comment]
me too

> > I believe this issue [d8ahazard/sd_dreambooth_extension#859](https://github.com/d8ahazard/sd_dreambooth_extension/issues/859) is related and I think you installed xformers==0.0.16 and you need to upgrade xformers such as
> > ```
> > pip install xformers==0.0.17.dev466
> > ```
> 
> Could not find a version that satisfies the requirement xformers==0.0.17.dev466 So I updated xformers=0.0.17.dev480, but it doesn't work for me. Still get NaN while training.

**Comment by dong-y21** [comment]
> > I believe this issue [d8ahazard/sd_dreambooth_extension#859](https://github.com/d8ahazard/sd_dreambooth_extension/issues/859) is related and I think you installed xformers==0.0.16 and you need to upgrade xformers such as
> > ```
> > pip install xformers==0.0.17.dev466
> > ```
> 
> Could not find a version that satisfies the requirement xformers==0.0.17.dev466 So I updated xformers=0.0.17.dev480, but it doesn't work for me. Still get NaN while training.

I updated xformers==0.0.17.dev466 but it doesn't work for me too. Still get NaN while training.

**Comment by mkshing** [comment]
Hi, I wonder if you still get Nan after removing `--use_8bit_adam`.

**Comment by dong-y21** [comment]
> Hi, I wonder if you still get Nan after removing `--use_8bit_adam`.

Unfortunately, it still turns into NaN after I removed '--use_8bit_adam'.
This is my command：
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch pretrain_e4t.py \
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" \
  --clip_model_name_or_path="openai/clip-vit-large-patch14" \
  --domain_class_token="face" \
  --placeholder_token="*s" \
  --prompt_template="a photo of {placeholder_token}" \
  --reg_lambda=0.01 \
  --output_dir="pretrained-face-5" \
  --train_image_dataset="FFHQ" \
  --iterable_dataset \
  --resolution=512 \
  --train_batch_size=4 \
  --learning_rate=1.6e-5 \
  --max_train_steps=30000 \
  --mixed_precision="fp16" \
  --enable_xformers_memory_efficient_attention

And this is my origin command：
CUDA_VISIBLE_DEVICES=0,1 accelerate launch pretrain_e4t.py \
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" \
  --clip_model_name_or_path="openai/clip-vit-large-patch14" \
  --domain_class_token="face" \
  --placeholder_token="*s" \
  --prompt_template="a photo of {placeholder_token}" \
  --reg_lambda=0.01 \
  --output_dir="pretrained-face-5" \
  --train_image_dataset="/share_graphics_ai/dongyi03/SIGGRAPH/controlled_experiment/dataset/FFHQ" \
  --iterable_dataset \
  --resolution=512 \
  --train_batch_size=8 \
  --learning_rate=1.6e-5 \
  --max_train_steps=30000 \
  --mixed_precision="fp16" \
  --enable_xformers_memory_efficient_attention \
  --use_8bit_adam

Since I have to reduce the batch_size on single V100 and use more GPU devices after removing '--use_8bit_adam'.
Yet these two commands both become NaN around 10k steps.
<img width="536" alt="image" src="https://user-images.githubusercontent.com/106572559/227774013-2d7ba4ec-f1e5-4861-9476-8c944ca8c8eb.png">

**Comment by dong-y21** [comment]
> Hi, I wonder if you still get Nan after removing `--use_8bit_adam`.

Unfortunately, it still turns into NaN after I removed '--use_8bit_adam'.
This is my command：
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch pretrain_e4t.py
--pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5"
--clip_model_name_or_path="openai/clip-vit-large-patch14"
--domain_class_token="face"
--placeholder_token="*s"
--prompt_template="a photo of {placeholder_token}"
--reg_lambda=0.01
--output_dir="pretrained-face-5"
--train_image_dataset="FFHQ"
--iterable_dataset
--resolution=512
--train_batch_size=4
--learning_rate=1.6e-5
--max_train_steps=30000
--mixed_precision="fp16"
--enable_xformers_memory_efficient_attention

And this is my origin command：
CUDA_VISIBLE_DEVICES=0,1 accelerate launch pretrain_e4t.py
--pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5"
--clip_model_name_or_path="openai/clip-vit-large-patch14"
--domain_class_token="face"
--placeholder_token="*s"
--prompt_template="a photo of {placeholder_token}"
--reg_lambda=0.01
--output_dir="pretrained-face-4"
--train_image_dataset="FFHQ"
--iterable_dataset
--resolution=512
--train_batch_size=8
--learning_rate=1.6e-5
--max_train_steps=30000
--mixed_precision="fp16"
--enable_xformers_memory_efficient_attention
--use_8bit_adam

Since I have to reduce the batch_size on single V100 and use more GPU devices after removing '--use_8bit_adam'.
Yet these two commands both become NaN around 10k steps.
<img width="536" alt="截屏2023-03-26 19 56 52" src="https://user-images.githubusercontent.com/106572559/227774117-fcfc2a89-a19b-41a0-85d0-74fd3d81e248.png">

**Comment by Actis92** [comment]
I had the same error, and the only way that has worked for me is to use float32 precision during the pretraining

**Comment by dong-y21** [comment]
> I had the same error, and the only way that has worked for me is to use float32 precision during the pretraining

But my loss will diverge quickly and still turn to NaN if I change '--mixed_precision' from 'fp16' to 'no'. 
I wonder whether this code can work on multi GPU devices.
<img width="469" alt="image" src="https://user-images.githubusercontent.com/106572559/227862498-f4fad23a-b490-464f-a4b3-dd4b280d5535.png">
This is my command:
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch pretrain_e4t.py \
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" \
  --clip_model_name_or_path="openai/clip-vit-large-patch14" \
  --domain_class_token="face" \
  --placeholder_token="*s" \
  --prompt_template="a photo of {placeholder_token}" \
  --reg_lambda=0.01 \
  --output_dir="pretrained-face-2" \
  --train_image_dataset="FFHQ" \
  --iterable_dataset \
  --resolution=512 \
  --train_batch_size=4 \
  --learning_rate=1.6e-5 \
  --max_train_steps=30000 \
  --mixed_precision="no" \
  --enable_xformers_memory_efficient_attention \
  --use_8bit_adam

**Comment by dong-y21** [comment]
> What I have tried is mixed_precision="no" but also without use_8bit_adam, I'm also using two gpus 😅 I don't remember if I have done other changes in the code that could have an impact on this problem

Thanks, I will try your training config. And I also wonder if you have changed --train_batch_size or --learning_rate.

**Comment by Actis92** [comment]
> > What I have tried is mixed_precision="no" but also without use_8bit_adam, I'm also using two gpus 😅 I don't remember if I have done other changes in the code that could have an impact on this problem
> 
> Thanks, I will try your training config. And I also wonder if you have changed --train_batch_size or --learning_rate.

I'm using train_batch_size=8 since using 2 machines gives a real batch of 16 (as in the paper) and learning_rate=1.6e-5 (the default)

**Comment by dong-y21** [comment]
> What I have tried is mixed_precision="no" but also without use_8bit_adam, I'm also using two gpus 😅 I don't remember if I have done other changes in the code that could have an impact on this problem

This config really works! 

CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch pretrain_e4t.py \
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" \
  --clip_model_name_or_path="openai/clip-vit-large-patch14" \
  --domain_class_token="face" \
  --placeholder_token="*s" \
  --prompt_template="a photo of {placeholder_token}" \
  --reg_lambda=0.01 \
  --output_dir="pretrained-face-2" \
  --train_image_dataset="FFHQ" \
  --iterable_dataset \
  --resolution=512 \
  --train_batch_size=4 \
  --learning_rate=1.6e-5 \
  --max_train_steps=30000 \
  --mixed_precision="no" \
  --enable_xformers_memory_efficient_attention

**Comment by mkshing** [comment]
I changed a lot of parts to get better results and released the pre-trained model for face. You should check the latest code and reopen this issue if you have still problems.
